src/core/server/http/router/validator.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SchemaTypeError, ValidationError, ObjectType, Type, TypeOf } from '@kbn/config-schema';
+
+/**
+ * Allowed returned format of the custom validate function
+ * @public
+ */
+export type ValidateFunctionReturn<T> =
+  | {
+      value: T;
+      error?: undefined;
+    }
+  | {
+      value?: undefined;
+      error: SchemaTypeError;
+    };
+
+/**
+ * Custom validate function (only if @kbn/config-schema is not a valid option in your plugin)
+ * @public
+ */
+export type ValidateFunction<T> = (data: any) => ValidateFunctionReturn<T>;
+
+export type TypeOfFunctionReturn<T extends ValidateFunction<unknown>> = NonNullable<
+  ReturnType<T>['value']
+>;
+
+export type ValidateSpecs<T> = ObjectType | Type<T> | ValidateFunction<T>;
+export type ValidatedType<T extends ValidateSpecs<unknown>> = T extends Type<unknown>
+  ? TypeOf<T>
+  : T extends ValidateFunction<unknown>
+  ? TypeOfFunctionReturn<T>
+  : never;
+
+function validateFunction<T extends ValidateFunction<unknown>>(
+  validationSpec: T,
+  data: any,
+  namespace?: string
+): TypeOfFunctionReturn<typeof validationSpec> {
+  let result: ValidateFunctionReturn<ReturnType<typeof validationSpec>['value']>;
+  try {
+    result = validationSpec(data);
+  } catch (err) {
+    result = { error: new SchemaTypeError(err, []) };
```

Comment:
I don't follow what a desirable way to fail validation is: to throw an exception or return an exception?

Reply:
The desirable way is to return `{ value: validatedValue }` or `{ error: RouteValidationError }`. But since we don't have control over the validation method and it can throw an error, we don't want to break the process and throw a 500 error (or do we?).